core/include/gap/core/optional.hpp

```diff
@@ -0,0 +1,225 @@
+        constexpr value_type&& value() && { return std::move(**this); }
+        constexpr const value_type&& value() const&& { return std::move(**this); }
+
+        template< typename U >
+            requires(copy_constructible && std::is_convertible_v< U&&, value_type >)
+        constexpr value_type value_or( U&& default_value ) const& {
+            return bool(*this) ? **this : static_cast< value_type >(std::forward<U>(default_value));
+        }
+
+        template< typename U >
+            requires(move_constructible && std::is_convertible_v< U&&, value_type >)
+        constexpr value_type value_or( U&& default_value ) && {
+            return bool(*this) ? std::move(**this) : static_cast< value_type >(std::forward<U>(default_value));
+        }
+    };
+
+    //
+    // optional wrapper
+    //
+    namespace detail {
+        template< typename T >
+        using optional_wrapper_base = std::conditional_t<
+            optional_like< T >, T, optional_adaptor< T >
+        >;
+    } // namespace detail
+
+    template< typename base >
+    struct optional_wrapper : detail::optional_wrapper_base< base >
+    {
+        using base_type  = detail::optional_wrapper_base< base >;
+        using value_type = typename base_type::value_type;
+
+        using base_type::base_type;
+        using base_type::value;
+
+        static constexpr bool copy_constructible = std::is_copy_constructible_v< base_type >;
+        static constexpr bool move_constructible = std::is_copy_constructible_v< base_type >;
+
+        static constexpr bool nothrow_move_constructible
+            = std::is_nothrow_move_constructible_v< base_type >;
+
+        constexpr optional_wrapper() noexcept = default;
+
+        constexpr optional_wrapper(const base_type &other)
+            requires(copy_constructible)
+            : base_type(other) {}
+
+        constexpr optional_wrapper(base_type &&other)
+            noexcept(nothrow_move_constructible)
+            requires(move_constructible)
+            : base_type(std::move(other))
+        {}
+
+        template< typename U >
+        requires(std::is_constructible_v< base_type, optional_wrapper< U > && >)
+        explicit(!std::is_convertible_v< const U&, value_type >)
+        constexpr optional_wrapper(optional_wrapper< U > &&other)
+            : base_type(std::move(other))
+        {}
+
+        template< typename U >
+        requires(std::is_constructible_v< base_type, const optional_wrapper< U > & >)
+        explicit(!std::is_convertible_v< U&&, value_type >)
+        constexpr optional_wrapper(const optional_wrapper< U > &other)
+            : base_type(other)
+        {}
+
+        template< typename U = value_type >
+        requires(std::is_constructible_v< base_type, U&& >)
+        explicit(!std::is_convertible_v< U&&, value_type >)
+        constexpr optional_wrapper( U &&other )
+            : base_type(std::forward< U >(other))
+        {}
+
+        base& unwrap() & { return *this; }
```

Comment:
Why is this not `value()` from the parent class?

Reply:
it is completely different thing :P `value` returns `**this` stored value, whereas `unwrap` casts to base class `*this`